File: bun-hono/src/modules/contact/contact.service.ts
import ContactRepository from "./contact.respository";
import Contact from "./contact";
import AppError from "../../common/api.error";
import ILogger from "../../config/logger/Ilogger";

class ContactService {
  
  constructor(private contactRepository: ContactRepository, private logger: ILogger) {}

  async createContact(name: string, email: string, phone: string, userId: string): Promise<Contact> {
    
    this.logger.debug("Service Contact create Called")
    const resultContact = await this.contactRepository.create(name, email, phone, userId)
    return resultContact;
   
  }

  async findAllContact(userId: string): Promise<Contact[]>{

    this.logger.debug("Service Contact findAll Called", { userId })
    const contacts = await this.contactRepository.findAll(userId)
    return contacts

  }

  async updateContact(id: string, userId: string, contact: Partial<Contact>) : Promise<Partial<Contact>> {

    const existingContact = await this.contactRepository.findByID(id)
    if (!existingContact) {
      throw new AppError(404, "Contact not found")
    }
    if (existingContact.userId !== userId) {
      throw new AppError(403, "Forbidden")
    }
    const updatedContact = await this.contactRepository.update(id, contact)
    return updatedContact

  }

  async deleteContact(id: string, userId: string) : Promise<Partial<Contact>> {

    const existingContact = await this.contactRepository.findByID(id)
    if(!existingContact){
      throw new AppError(404, "Contact not found")
    }
    if (existingContact.userId !== userId){
      throw new AppError(403, "Forbidden")
    }
    const deletedContact = await this.contactRepository.delete(id)
    return deletedContact

  }
}

export default ContactService;

File: spring-java/src/main/java/org/distr/rcontacts/service/ContactService.java
package org.distr.rcontacts.service;

import java.util.List;

import org.distr.rcontacts.repository.sql.ContactRepository;
import org.distr.rcontacts.entities.ContactEntity;
import org.distr.rcontacts.exception.AppException;
import org.distr.rcontacts.contracts.ContactContract;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ContactService {
    
    private ContactRepository contactRepository;

    public ContactService(ContactRepository contactRepository) {
        this.contactRepository = contactRepository;
    }

    public String test() {
        return "Hello from ContactService!";
    }

    @Transactional
    public ContactContract CreateContact(String userId, String name, String phone, String email){
        ContactEntity contact = new ContactEntity(
            null,
            userId,
            name,
            phone,
            email
        );
        var result = contactRepository.create(contact);
        var contactContract = new ContactContract(
            result.getId(),
            result.getUserId(),
            result.getName(),
            result.getPhone(),
            result.getEmail()
        );
        return contactContract;
        
    }

    @Transactional( readOnly = true )
    public List<ContactContract> getAllContactsByUserId(String userId){
        var contacts = contactRepository.findAllByUserId(userId);
        return contacts.stream().map(contact -> new ContactContract(
            contact.getId(),
            contact.getUserId(),
            contact.getName(),
            contact.getPhone(),
            contact.getEmail()
        )).toList();
    }

    @Transactional
    public void deleteContact(String contactId){
        var existingContact = contactRepository.findById(contactId);
        if(existingContact.isEmpty()){
            throw new AppException("Contact not Found", 404);
        }
        contactRepository.delete(contactId);
    }

    @Transactional
    public ContactContract updateContact(String contactId, String userId,  String name, String phone, String email){
        ContactEntity contact = new ContactEntity(
            contactId,
            userId,
            name,
            phone,
            email
        );
        var existingContact = contactRepository.findById(contactId);
        if(existingContact.isEmpty()){
            throw new AppException("Contact not Found", 404);
        }
        if(!existingContact.get().getUserId().equals(userId)){
            throw new AppException("Unauthorized", 403);
        }
        var result = contactRepository.update(contact);
        var contactContract = new ContactContract(
            result.getId(),
            result.getUserId(),
            result.getName(),
            result.getPhone(),
            result.getEmail()
        );
        return contactContract;
    }

}
